### ISSUE TITLE
Middleware added Nonce attributes missing from TypeScript production builds

### ISSUE THREAD

**Issue by andymagill** [vercel/next.js#66871]
### Link to the code that reproduces this issue

https://github.com/andymagill/csp-middleware-nextjs-bug/tree/main

### To Reproduce

1. Create a local copy of the reproduction template via `yarn create next-app --example reproduction-template reproduction-app`
2. Add the [middleware.ts](https://gist.github.com/andymagill/678ab6ac90e616f93fd57a8f409ec8fa) TypeScript (not JS) file from the documentation : [https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy](https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy)
3. Build the project via `yarn install` then `yarn run build` then `yarn run start`
4. Open the browser console to see CSP errors for each bundle. Alternatively, inspect the source of the server response to confirm bundles have no nonce attribute. 

### Current vs. Expected behavior

**Current behavior**:  nonce attributes are missing from bundles
**Expected behavior**: nonce attributes are present on bundles

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 11 Pro
  Available memory (MB): 16093
  Available CPU cores: 16
Binaries:
  Node: 20.11.0
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.2.4 // Latest available version is detected (14.2.4).
  eslint-config-next: N/A
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 4.9.5
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Documentation, Middleware, TypeScript

### Which stage(s) are affected? (Select all that apply)

next build (local)

### Additional context

This behavior appears to be specific to _production_ builds of next.js _typescript_ _middleware_. Running `yarn run dev` in the example repo works fine and correctly adds the nonce.  The "with-strict-csp" example with the middleware.JS (not ts) also correctly adds the nonce on production builds.

Currently, I see no way to add nonces to Next.JS TypeScript projects using the documented methods.

**Comment by ebCrypto** [comment]
Thank you @andymagill for posting this. I feel like I am encountering the same issue but with js instead of ts and the latest version of next.
Have you found a workaround by any chance?

**Comment by terrymun** [comment]
We also ran into this peculiar bug: for some reason, doing a noop call to `headers()` on the problematic route's layout.tsx somehow fixes the issue for us:

```tsx
const Layout = ({ children }: PropsWithChildren) => {
  // HACK: Mysterious bug in next.js, where manually calling header is needed to force nonce injecton on the page when building
  headers().get('x-nonce');

  return (
    // Some TSX stuff here
  );
};

export default Layout;
```

**Comment by darushton** [comment]
We also had the exact issue described - can be reproduced using `npx create-next-app@latest` and the nonce middleware example from https://nextjs.org/docs/app/guides/content-security-policy.

@terrymun - great spot, that works for us too.

@samcx - are there any plans for an official fix for this please?

**Comment by samcx** [comment]
@darushton Pinged the team about it!

**Comment by icyJoseph** [comment]
Hi,

Is the issue here that `nonce` necessarily requires dynamic pages? We do mention this in https://nextjs.org/docs/app/guides/content-security-policy#adding-a-nonce-with-middleware, although recently we also got feedback that it is not clear enough:

> Every time a page is viewed, a fresh nonce should be generated. This means that you must use dynamic rendering to add nonces.

The pre-rendered pages, don't have the `nonce` generated by the middleware. That's why adding things like `headers()` and such, that opt into dynamic behavior does the trick. 

We also have an experimental `sri` option for Webpack:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  experimental: {
    sri: {
      algorithm: 'sha256', // or 'sha384' or 'sha512'
    },
  },
}

module.exports = nextConfig
```

That'll add some content based hashes - and such. We are working to improve the CSP documentation. 

Note that nonces and dynamic behavior are not a Next.js blocker. https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CSP#nonces also says:

- In practice this means that the nonce must be different for every HTTP response, and must not be predictable.

A fair counter point, would be, why can't Next.js load into memory the HTML is about to send down, even if pre-rendered, and sprinkle the `nonce` to scripts, but that'd cause hydration errors, and be a kind of implicit dynamic behavior added to an otherwise static build. I don't even know if this can be implemented, I just thought I mention it for completeness, there's probably other implications I am missing.

**Comment by terrymun** [comment]
@icyJoseph It is an implicit assumption that when a user knows what to do when you see "This means that you must use dynamic rendering to add nonces". I suppose it would help many users, especially those who are relatively new to next.js, to have explicit instruction on the nonce page to ensure that the route is opted in to dynamic rendering for nonce to work 😃

**Comment by icyJoseph** [comment]
Yeah for sure, like I said we are working to improve this document. I am a bit worried, cuz knowing what's dynamic rendering is important in a PPR world. https://nextjs.org/docs/app/getting-started/partial-prerendering#dynamic-rendering

This feedback is important though. As far as this issue goes though, we can maybe change it to a documentation issue, if not close?

**Comment by darushton** [comment]
Thanks for your detailed response, @icyJoseph. That makes a lot of sense. Slightly embarrassed now reviewing the doc and seeing "you must use dynamic rendering" highlighted in bold 😅

Agree that changing this to a documentation issue rather than a bug sounds appropriate - even just a link off to the dynamic rendering section of the PPR page would be helpful to those new to it.

**Comment by terrymun** [comment]
@icyJoseph I think just minor changes to the docs as it is today will make a huge difference :)

- For the text "Every time a page is viewed, a fresh nonce should be generated. This means that you must use dynamic rendering to add nonces.", it would be so helpful to add the [link to the dynamic rendering page](https://nextjs.org/docs/app/getting-started/partial-prerendering#dynamic-rendering)
- We could benefit from a code snippet where a root `layout.tsx` is involved. There are many cases where the root `layout.tsx` of a large project actually does not consume the `nonce` at all, but as all scripts and stylesheets are injected at this level it is needed to be opted into dynamic rendering.

**Comment by icyJoseph** [comment]
I'll do the first part as part of this issue. We have an open PR, worked by Lee , Devin and my feedback. I'll take over that one, and make some adjustments. That PR already contains an example with layouts. It just needs more time, and point 1 here is rather straight forward.
